### PR TITLE
codeintel: expose code intel info for `GitBlob` 

### PR DESCRIPTION
**Pull request by Strum355**
Closes https://github.com/sourcegraph/sourcegraph/issues/31073
Reliant on https://github.com/sourcegraph/go-ctags/pull/7

## Test plan

Tested by hand in the API console. The brunt of the work is done in symbols service/go-ctags, and we dont have a great way of testing symbols service interfacing yet, sot his has been mostly manually tested

### CODE REVIEW

File: cmd/symbols/main.go [hunk 1]
```diff
@@ -104,15 +104,15 @@ func main() {
 
 	parserPool, err := parser.NewParserPool(ctagsParserFactory, config.numCtagsProcesses)
 	if err != nil {
-		log.Fatalf("Failed to parser pool: %s", err)
+		log.Fatalf("Failed to create parser pool: %s", err)
```
Comment (efritz) [inline cmd/symbols/main.go:107]: _failed to parser pool_ 👨‍🍳

File: cmd/symbols/internal/api/handler.go [hunk 2]
```diff
@@ -70,6 +77,17 @@ func (h *apiHandler) handleSearch(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func (h *apiHandler) handleListLanguages(w http.ResponseWriter, r *http.Request) {
+	mapping, err := ctags.ListLanguageMappings(r.Context(), h.ctagsBinary)
```
Comment (efritz) [inline cmd/symbols/internal/api/handler.go:81]: You gotta update ctags dep here?

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+			compiled, err := glob.Compile(pattern)
+			if err != nil {
+				return nil, err
+			}
+
+			if compiled.Match(path.Base(r.gitBlobMeta.Path)) {
+				return NewSearchBasedCodeIntelResolver(&allowedLanguage), nil
+			}
+		}
+	}
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go:178]: I feel like this is a good candidate for something in the lower resolver layer. Would be nice to have this tested and callable outside of graphql.

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+func (r *searchBasedSupportResolver) Language() *string {
+	return r.language
+}
+
+type preciseCodeIntelSupportResolver struct {
+	indexers []gql.CodeIntelIndexerResolver
+}
+
+func NewPreciseCodeIntelSupportResolver(filepath string) gql.PreciseCodeIntelSupportResolver {
+	fmt.Println(path.Ext(filepath))
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go:211]:
```suggestion
```

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+
+type preciseCodeIntelSupportType string
+
+const (
+	native     preciseCodeIntelSupportType = "NATIVE"
+	thirdParty preciseCodeIntelSupportType = "THIRD_PARTY"
+	unknown    preciseCodeIntelSupportType = "UNKNOWN"
+)
+
+var (
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go:32]: Can we move these declarations out into some other const-like package? This should be a nice source-of-truth for things like auto-generated docs/directories in the future.

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+		errTracer:   errTracer,
+	}
+}
+
+func (r *gitBlobCodeIntelInfoResolver) Support(ctx context.Context) gql.CodeIntelSupportResolver {
+	return NewCodeIntelSupportResolver(r.gitBlobMeta, r.errTracer)
+}
+
+func (r *gitBlobCodeIntelInfoResolver) LSIFUploads(ctx context.Context) (gql.LSIFUploadConnectionResolver, error) {
+	return nil, nil
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go:144]: TODO?

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+	if hasNative {
+		return string(native)
+	} else if len(r.indexers) > 0 {
+		return string(thirdParty)
+	} else {
+		return string(unknown)
+	}
+}
+
+func (r *preciseCodeIntelSupportResolver) Indexers() *[]gql.CodeIntelIndexerResolver {
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go:235]: Does this have to be a pointer?

**Review by efritz — CHANGES_REQUESTED** [review]

File: cmd/symbols/internal/api/handler.go (hunk 2 above)
Comment (Strum355) [reply cmd/symbols/internal/api/handler.go:81]: Yea, that PR is pending

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go (hunk 4 above)
Comment (efritz) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go:211]: I'm gonna tweet charity majors

File: cmd/symbols/internal/api/handler.go (hunk 2 above)
Comment (efritz) [reply cmd/symbols/internal/api/handler.go:81]: - [ ] update go.mod

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	for _, allowedLanguage := range ctags.SupportedLanguages {
+		for _, pattern := range mappings[allowedLanguage] {
+			compiled, err := glob.Compile(pattern)
+			if err != nil {
+				return nil, err
+			}
```
Comment (mrnugget) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go:172]: It probably doesn't matter, but I'd turn the loop around and compile the patterns only once.

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+	return &preciseCodeIntelSupportResolver{
+		indexers: languageToIndexer[path.Ext(filepath)],
+	}
+}
+
+func (r *preciseCodeIntelSupportResolver) SupportLevel() string {
+	var hasNative bool
+	for _, indexer := range r.indexers {
+		if strings.HasPrefix(indexer.URL(), "https://github.com/sourcegraph") {
+			hasNative = true
```
Comment (mrnugget) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go:221]: This feels like something that could be defined on the indexers and not in the GraphQL layer. As in: a `IsNative() bool` method on the indexers.

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go (hunk 7 above)
Comment (Strum355) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go:235]: the slice? Unfortunately yea, else it complains about it being "not a pointer" for purposes of being nil...even though a slice can be nil

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go (hunk 8 above)
Comment (Strum355) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go:172]: Not sure what way youre thinking, `glob.Compile` only takes a single pattern, and each language has a unique (as i understand) set of patterns

Comment (mrnugget) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go:172]: Oh, I see. I misread the loop the first time.

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go (hunk 6 above)
Comment (Strum355) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go:143]: Done in [`430b957` (#31442)](https://github.com/sourcegraph/sourcegraph/pull/31442/commits/430b9573f57b08ae25680b71a687f99639ee2722)

File: enterprise/cmd/frontend/internal/codeintel/resolvers/query_uploads.go [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/codeintel/stores/dbstore"
+)
+
+// LSIFUploads returns the list of dbstore.Uploads for the store.Dumps determined to be applicable
+// for answering code-intel queries.
+func (r *queryResolver) LSIFUploads(ctx context.Context) (uploads []dbstore.Upload, err error) {
+	return r.dbStore.GetUploadsByIDs(ctx, r.closestDumpIDs()...)
+}
+
+func (r *queryResolver) closestDumpIDs() []int {
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/query_uploads.go:15]: :laff: you actually did it you madman

Comment (Strum355) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/query_uploads.go:15]: image

File: cmd/frontend/graphqlbackend/codeintel.graphql [hunk 11]
```diff
@@ -1332,3 +1334,41 @@ type IndexConfiguration {
     """
     inferredConfiguration: String
 }
+
+type GitBlobCodeIntelInfo {
```
[the remark this answers is not in the document]
Comment (efritz) [reply cmd/frontend/graphqlbackend/codeintel.graphql:1353]: Plan to still do this pre-merge?

[the remark this answers is not in the document]
Comment (Strum355) [reply cmd/frontend/graphqlbackend/codeintel.graphql:1353]: Yea in progress rn

[the remark this answers is not in the document]
Comment (Strum355) [reply cmd/frontend/graphqlbackend/codeintel.graphql]: Done here, lmk wyt [`e8ee841` (#31442)](https://github.com/sourcegraph/sourcegraph/pull/31442/commits/e8ee8418c9f42fd61f8b2d35522969592edc55f8)

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go (hunk 5 above)
Comment (Strum355) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go]: Done here, lmk wyt [`44e5cd8` (#31442)](https://github.com/sourcegraph/sourcegraph/pull/31442/commits/44e5cd84f402102318f18e9c7af17cc7419bbe76)

**Review by efritz — APPROVED** [review]
Nothing blocking :)

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go (hunk 9 above)
Comment (Strum355) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go]: I've changed this code for a new invariant, so I dont know if this is still particularly relevant

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go (hunk 3 above)
Comment (Strum355) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go]: Done here, lmk wyt etc etc [`ab3a4f0` (#31442)](https://github.com/sourcegraph/sourcegraph/pull/31442/commits/ab3a4f00597f5a00b1e8b47c1b0c749a9c99005b)

Comment (efritz) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/codeintel_info_resolver.go]: :shipit:

**Pull request merged**